docs/node/mainnet: Add 23.0 upgrade instructions for the Mainnet (#577)
oasisprotocol/docs: docs/node/mainnet/e-upgrade.md

:::

[Testnet upgrade]: ../testnet/upgrade-log.md

### Voting

On XXX, an upgrade proposal will be proposed which (if accepted) will schedule
the upgrade on epoch **XXX**.

For an optimized voting experience, we recommend utilizing the [Oasis CLI].
Follow these steps to cast your vote:
1. [Import your keys into the wallet]
2. [Cast your vote]

### Upgrade Instructions

The following steps should be performed only after the network has reached the
upgrade epoch and has halted:

1. Wait for the network to reach the upgrade epoch and halt. The node
    will automatically stop without any action required on your part. After the
    network has halted, proceed to the next steps.

2. Download the Mainnet genesis file published in the
   [Mainnet XXX release].

:::info

Mainnet state at epoch **XXX** will be exported and migrated to a 23.0
compatible genesis file.

The new genesis file will be published on the above link soon after reaching the
upgrade epoch.

:::

3. Verify the provided Mainnet upgrade genesis file by comparing it to the
   local network state dump.

>    Find the `genesis-mainnet-XXX.json` file in the `exports` subdirectory in your data dir (e.g. `/node/`, `/srv/oasis/node/`)  and run

Reviewer: Line seems too long and should be wrapped.

>    sha256sum on it. Afterwards, compare it with the hash that we will share on

Reviewer: ```suggestion
   `sha256sum` on it. Afterwards, compare it with the hash that we will share on
```

   the `#node-operators` Discord channel.

   The state changes are described in the [State Changes](#state-changes)
   section below.

4. Replace the old genesis file with the new Mainnet genesis file.

5. Ensure your node will remain stopped by disabling auto-starting via your
   process manager (e.g., [systemd] or [Supervisor])

6. Back up the entire data directory of your node. Verify that the backup
includes the following folders:
- for consensus: `tendermint/abci-state` and `tendermint/data`
- for runtimes: `runtimes/*/mkvs_storage.badger.db` and
`runtimes/*/worker-local-storage.badger.db`

7. [Wipe state]. This must be performed _before_ replacing the Oasis Node
   binary. In case you are upgrading ParaTimes/runtimes ensure you read the

>    following Danger section:

Reviewer: This is a bit strange as the section is right here? Will people really not read this if not pointed to it?

Author: We had a lot of questions about this during testnet upgrade even thou danger section is right here. This is why I would point it out. Maybe people jumped over the Danger section.

Reviewer: Ok, then maybe just:
```suggestion
   binary. In case you are upgrading ParaTimes/runtimes ensure you read the
   following section:
```

:::danger

State of ParaTimes/runtimes is not affected by this upgrade and MUST NOT be
wiped. Wiping state for confidential ParaTimes will prevent your compute or
key manager node from transitioning to the new network.

Make sure to include the `--preserve.mkvs_database` flag if you are wiping

> blockchain state on a runtime node and need to preserve runtime state.

Reviewer: This is not needed as this flag is the default.

Author: We had a lot of questions about this flag. Also https://docs.oasis.io/node/run-your-node/maintenance/wiping-node-state#state-wipe-and-keep-node-identity does not look like it is default?

Reviewer: It sure is:
```
$ ./bin/oasis-node unsafe-reset --help
reset the node state (UNSAFE)

Usage:
  oasis-node unsafe-reset [flags]

Flags:
  -n, --dry_run                  don't actually do anything, just show what will be done
      --force                    force
  -h, --help                     help for unsafe-reset
      --preserve.local_storage   preserve per-runtime untrusted local storage (default true)
      --preserve.mkvs_database   preserve per-runtime MKVS database (default true)
...
```

Reviewer: The linked doc seems out of date, this has been the default for a long time now. The doc should be fixed instead and this being the default should be pointed out here (without mentioning the flag).

Author: Will edit the linked docs after we finish this docs.

Author: Oh something crashed. Let me check what it is

Transitioning confidential ParaTimes to the new network requires local state
that is sealed to the CPU. This also means that bootstrapping a new node on a
separate CPU immediately after the network upgrade will not be possible until
an updated ParaTime containing new trust roots is released and adopted.

:::

> 8. Replace the old version of Oasis Node with version [23.0].

Reviewer: Version will be at least 23.0.2.

Author: I changed it to 23.0.x. We can add the final version later.

Note that the migration subcommand does not preserve comments and order of
sections from the input YAML config file. You should always carefully read the
output of this command, as well as compare the generated config file with the
original before using it.

After you are satisfied with the new configuration file, replace the old file
with the new one as follows:

```
mv new-config.yml config.yml
```

:::tip

The configuration format for seed nodes has changed and it now requires the
node's P2P public key to be used. In case your old configuration file contains
known Mainnet seed nodes, this transformation is performed automatically.

However, if it contains unknown seed nodes then the conversion did not happen
automatically and you may need to obtain the seed node's P2P public key. For
Mainnet you can use the following addresses:

* `XXX`
* `XXX`

Please be aware that every seed node should be configured to listen on two
distinct ports. One is dedicated to peer discovery within the CometBFT P2P
network, while the other is used to bootstrap the Oasis P2P network.

:::

[Change Log]:
  https://github.com/oasisprotocol/oasis-core/blob/v23.0/CHANGELOG.md

>   ### Data Directory Changes

Reviewer: ```suggestion
### Data Directory Changes
```

The subdirectory (located inside the node's data directory) used to store
consensus-related data, previously called `tendermint` (after the consensus
layer protocol backend) has been renamed to `consensus` in Oasis Core 23.0. If
any of your scripts rely on specific directory names, please make sure to update
them to reflect the changed sdirectory name.

### State Changes

The following parts of the genesis document will be updated:

:::info

For a more detailed explanation of the parameters below, see the
[Genesis Document] docs.

:::

:::info

All state changes will be done automatically with the migration command provided
by the new version of `oasis-node`. It can be used as follows to derive the same
genesis file from an existing state dump at the correct height (assuming there
is a `genesis.json` present in the current working directory):

```

> oasis-node genesis migrate --genesis.new_chain_id XXX

Reviewer: ```suggestion
oasis-node genesis migrate --genesis.new_chain_id oasis-4
```

```

:::

#### General

> * **`chain_id`** will be set to `XXX`.

Reviewer: ```suggestion
* **`chain_id`** will be set to `oasis-4`.
```

6. Back up the entire data directory of your node. Verify that the backup
includes the following folders:
- for consensus: `tendermint/abci-state` and `tendermint/data`
- for runtimes: `runtimes/*/mkvs_storage.badger.db` and
`runtimes/*/worker-local-storage.badger.db`

7. [Wipe state]. This must be performed _before_ replacing the Oasis Node
   binary. In case you are upgrading ParaTimes/runtimes ensure you read the
   following Danger section:

:::danger

State of ParaTimes/runtimes is not affected by this upgrade and MUST NOT be
wiped. Wiping state for confidential ParaTimes will prevent your compute or
key manager node from transitioning to the new network.

To safely wipe the blockchain state on a runtime while preserving the runtime
state, follow these steps:
1. **Dry Run:** Initiate a dry run to preview which files will be deleted
by running the following command: `oasis-node unsafe-reset --dry_run`
2. **Wipe blockchain state:** After reviewing the dry run results, proceed with
the reset by running: `oasis-node unsafe-reset`

Note that by default, the `--preserve.mkvs_database` flag is set to true,
preserving the runtime state. You can find additional information about the

> `unsafe-reset` on the [Wipe state] page.

Reviewer: I would not include this, it seems confusing to mention a flag that defaults to true and is not necessary to be specified?

Author: During the Testnet upgrade, we told a lot of operators to include this flag, so it might be confusing to them that now there is no flag. I do not know what the best option is. Do I remove this?

Reviewer: Yes, I would remove this here and change the "wipe state" docs to specify that this flag is true by default.

:::

[Testnet upgrade]: ../testnet/upgrade-log.md

### Voting

On XXX, an upgrade proposal will be proposed which (if accepted) will schedule
the upgrade on epoch **XXX**.

> For an optimized voting experience, we recommend utilizing the [Oasis CLI].

Reviewer 2: ```suggestion
For optimal voting experience, we recommend using the [Oasis CLI].
```

Follow these steps to cast your vote:

> 1. [Import your keys into the wallet]

Reviewer: We should make sure to update the link once the new CLI release which will include easier file import. But this is probably fine for now.

upgrade epoch and has halted:

1. Wait for the network to reach the upgrade epoch and halt. The node
    will automatically stop without any action required on your part. After the
    network has halted, proceed to the next steps.

2. Download the Mainnet genesis file published in the
   [Mainnet XXX release].

:::info

Mainnet state at epoch **XXX** will be exported and migrated to a 23.0.x
compatible genesis file.

The new genesis file will be published on the above link soon after reaching the
upgrade epoch.

:::

3. Verify the provided Mainnet upgrade genesis file by comparing it to the
   local network state dump.

   Find the `genesis-mainnet-XXX.json` file in the `exports` subdirectory in
   your data dir (e.g. `/node/`, `/srv/oasis/node/`)  and run `sha256sum` on it.
   Afterwards, compare it with the hash that we will share on the
   `#node-operators` Discord channel.

   The state changes are described in the [State Changes](#state-changes)
   section below.

4. Replace the old genesis file with the new Mainnet genesis file.

5. Ensure your node will remain stopped by disabling auto-starting via your

>    process manager (e.g., [systemd] or [Supervisor])

Reviewer: ```suggestion
   process manager (e.g., [systemd] or [Supervisor]).
```

6. Back up the entire data directory of your node. Verify that the backup
includes the following folders:
- for consensus: `tendermint/abci-state` and `tendermint/data`
- for runtimes: `runtimes/*/mkvs_storage.badger.db` and
`runtimes/*/worker-local-storage.badger.db`

7. [Wipe state]. This must be performed _before_ replacing the Oasis Node
   binary. In case you are upgrading ParaTimes/runtimes ensure you read the
   following section:

:::danger

State of ParaTimes/runtimes is not affected by this upgrade and MUST NOT be
wiped. Wiping state for confidential ParaTimes will prevent your compute or
key manager node from transitioning to the new network.

To safely wipe the blockchain state on a runtime while preserving the runtime
state, follow these steps:

> 1. **Dry Run:** Initiate a dry run to preview which files will be deleted

Reviewer: ```suggestion
1. **Dry run:** initiate a dry run to preview which files will be deleted
```

by running the following command: 
```

> oasis-node unsafe-reset --dry_run

Reviewer: This is missing specifying the data directory? Or is the assumption that one runs this in the node's data directory -- in which case it should be explicitly stated.

Author: Added step 1: Go into your node's data directory

```

> 2. **Wipe blockchain state:** After reviewing the dry run results, proceed with

Reviewer: ```suggestion
2. **Wipe blockchain state:** after reviewing the dry run results, proceed with
```

the reset by running:
```

> oasis-node unsafe-reset

Reviewer: This is missing specifying the data directory?

Author: Added step 1: Go into your node's data directory

```

Transitioning confidential ParaTimes to the new network requires local state
that is sealed to the CPU. This also means that bootstrapping a new node on a
separate CPU immediately after the network upgrade will not be possible until
an updated ParaTime containing new trust roots is released and adopted.

:::

8. Replace the old version of Oasis Node with version [23.0.x].

:::info

The Oasis Core 23.0.x binary in our published releases is built only for Ubuntu
22.04 (GLIBC>=2.32). You'll have to build it yourself if you're using prior
Ubuntu versions (or other distributions using older system libraries).

:::

9. Perform any needed [configuration changes](#configuration-changes) described
   below.

10. (only Rosetta Gateway operators) Replace old version of Oasis Rosetta
    Gateway with version [2.6.0][rosetta-gw-2.6.0].

11. Start your node and re-enable auto-starting via your process manager.

[Oasis CLI]: ../../general/manage-tokens/cli/

> [Import your keys into the wallet]: ../../general/manage-tokens/cli/wallet/#import

Reviewer 2: This should be the right path once https://github.com/oasisprotocol/cli/pull/146 lands.

```suggestion
[Import your keys into the wallet]: ../../general/manage-tokens/cli/wallet.md#import-file
```

Author: Thanks! Link updated.

   `#node-operators` Discord channel.

   The state changes are described in the [State Changes](#state-changes)
   section below.

4. Replace the old genesis file with the new Mainnet genesis file.

5. Ensure your node will remain stopped by disabling auto-starting via your
   process manager (e.g., [systemd] or [Supervisor]).

6. Back up the entire data directory of your node. Verify that the backup
includes the following folders:
- for consensus: `tendermint/abci-state` and `tendermint/data`
- for runtimes: `runtimes/*/mkvs_storage.badger.db` and
`runtimes/*/worker-local-storage.badger.db`

7. [Wipe state]. This must be performed _before_ replacing the Oasis Node
   binary. In case you are upgrading ParaTimes/runtimes ensure you read the
   following section:

:::danger

State of ParaTimes/runtimes is not affected by this upgrade and MUST NOT be
wiped. Wiping state for confidential ParaTimes will prevent your compute or
key manager node from transitioning to the new network.

To safely wipe the blockchain state on a runtime while preserving the runtime
state, follow these steps:
1. Go into your node's data directory (e.g., `/serverdir/node`, `/node/`,

> `/srv/oasis/node/`)

Reviewer: ```suggestion
`/srv/oasis/node/`).
```

2. **Dry run:** initiate a dry run to preview which files will be deleted by
running the following command: 
```

> oasis-node unsafe-reset --dry_run

Reviewer: Have you tested that this works? If I don't specify the data directory I get:
```
level=error ts=2023-10-20T07:42:46.231628389Z caller=unsafe_reset.go:71 module=cmd/unsafe-reset msg="data directory must be set"
```

Author: I did not test this. Don't want to test this on my node, so it doesn't get accidentally wiped. I just assumed it would work. I changed the instructions now, to specify the data directory.
